File: mobile/lib/modules/incidents/services/incident_service.dart
/// Servicio de Incidentes Flutter (P2 — CU7-CU9).
library;

import 'dart:io';

import '../../../core/api_client.dart';
import '../models/incident_model.dart';

class IncidentService {
  // ── CU7 — Reportar incidente con fotos/audio/GPS ─────────────────

  static Future<IncidentDetail> reportIncident({
    required String titulo,
    required double latitud,
    required double longitud,
    String? descripcion,
    String? direccion,
    List<File>? fotos,
    File? audio,
  }) async {
    final fields = <String, String>{
      'titulo': titulo,
      'latitud': latitud.toString(),
      'longitud': longitud.toString(),
      ...?descripcion == null ? null : {'descripcion': descripcion},
      ...?direccion == null ? null : {'direccion': direccion},
    };

    final files = <MapEntry<String, File>>[];
    if (fotos != null) {
      for (final f in fotos.take(5)) {
        files.add(MapEntry('fotos', f));
      }
    }
    if (audio != null) {
      files.add(MapEntry('audio', audio));
    }

    return ApiClient.postMultipart<IncidentDetail>(
      '/incidents',
      fields: fields,
      files: files.isNotEmpty ? files : null,
      fromJson: (j) => IncidentDetail.fromJson(j as Map<String, dynamic>),
    );
  }

  // ── Listar mis incidentes ─────────────────────────────────────────

  static Future<List<Incident>> listMyIncidents() async {
    return ApiClient.get<List<Incident>>(
      '/incidents',
      fromJson: (j) => (j as List<dynamic>)
          .map((e) => Incident.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  // ── CU9 — Ficha técnica del incidente ────────────────────────────

  static Future<IncidentDetail> getDetail(int incidentId) async {
    return ApiClient.get<IncidentDetail>(
      '/incidents/$incidentId',
      fromJson: (j) => IncidentDetail.fromJson(j as Map<String, dynamic>),
    );
  }

  // ── CU8 — Re-clasificar con IA ───────────────────────────────────

  static Future<Classification> reclassify(int incidentId) async {
    return ApiClient.post<Classification>(
      '/incidents/$incidentId/classify',
      fromJson: (j) => Classification.fromJson(j as Map<String, dynamic>),
    );
  }
}

File: frontend/src/app/modules/p2_incidentes/incident.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environment';

// ── Interfaces ──

export interface IncidentMedia {
  id: number;
  incidente_id: number;
  tipo_medio: string;
  url_archivo: string;
  subido_en: string;
}

export interface Classification {
  id: number;
  incidente_id: number;
  categoria: string;
  severidad: string;
  confianza: number;
  razonamiento: string | null;
  metodo: string;
  clasificado_en: string;
}

export interface Incident {
  id: number;
  usuario_id: number;
  titulo: string;
  descripcion: string | null;
  estado: string;
  latitud: number;
  longitud: number;
  direccion: string | null;
  url_audio: string | null;
  severidad: string | null;
  categoria: string | null;
  creado_en: string;
  actualizado_en: string | null;
}

export interface IncidentDetail extends Incident {
  medios: IncidentMedia[];
  clasificacion: Classification | null;
}

@Injectable({ providedIn: 'root' })
export class IncidentService {
  private readonly http = inject(HttpClient);
  private readonly base = environment.apiUrl;

  /**
   * CU7 — Reportar incidente con fotos, audio y GPS.
   * Usa FormData para envío multipart.
   */
  reportIncident(
    data: { titulo: string; descripcion?: string; latitud: number; longitud: number; direccion?: string },
    fotos?: File[],
    audio?: File,
  ): Observable<IncidentDetail> {
    const fd = new FormData();
    fd.append('titulo', data.titulo);
    fd.append('latitud', data.latitud.toString());
    fd.append('longitud', data.longitud.toString());
    if (data.descripcion) fd.append('descripcion', data.descripcion);
    if (data.direccion) fd.append('direccion', data.direccion);
    if (fotos) fotos.forEach((f) => fd.append('fotos', f));
    if (audio) fd.append('audio', audio);
    return this.http.post<IncidentDetail>(`${this.base}/incidents`, fd);
  }

  /** Listar mis incidentes. */
  listMyIncidents(): Observable<Incident[]> {
    return this.http.get<Incident[]>(`${this.base}/incidents`);
  }

  /** Listar todos los incidentes (admin). */
  listAllIncidents(): Observable<Incident[]> {
    return this.http.get<Incident[]>(`${this.base}/incidents/all`);
  }

  /** CU9 — Ficha técnica completa. */
  getDetail(incidentId: number): Observable<IncidentDetail> {
    return this.http.get<IncidentDetail>(`${this.base}/incidents/${incidentId}`);
  }

  /** CU8 — Re-clasificar con IA. */
  reclassify(incidentId: number): Observable<Classification> {
    return this.http.post<Classification>(`${this.base}/incidents/${incidentId}/classify`, {});
  }
}
